components/suggestion/search_suggestion_list.jsx

```diff
@@ -29,6 +29,30 @@ export default class SearchSuggestionList extends SuggestionList {
         }
     }
 
+    generateLabel(item) {
+        if (item.username) {
```

Comment:
I missed how we were doing this before, but the `SuggestionList` and `SearchSuggestionList` shouldn't care what the type of their content is. The label should be determined by the various `Provider` classes since that lets us add a new `Provider` in the future without making any changes to `SuggestionList`.

For example, the way I'd add that to `SearchUserProvider` is to add a `generateLabel` method to it like
```javascript
generateLabel(item) {
    let label = item.username;
    if ((item.first_name || item.last_name) && item.nickname) {
        label += ` ${item.first_name} ${item.last_name} ${item.nickname}`;
    } else if (item.nickname) {
        label += ` ${item.nickname}`;
    } else if (item.first_name || item.last_name) {
        label += ` ${item.first_name} ${item.last_name}`;
    }
    return label;
}
```
and then modify the part of `handlePretextChanged` that calls `resultsCallback` to
```javascript
resultsCallback({
    matchedPretext: usernamePrefix,
    terms: mentions,
    items: users,
    component: SearchUserSuggestion,
    generateLabel: this.generateLabel,
});
```
You might also need to make `SuggestionBox` the one responsible for rendering the label, but that'll also save having to duplicate this between `SuggestionList` and `SearchSuggestionList`

Reply:
Never mind all this. While I'd like to do it more like this since it keeps the Providers being the ones responsible for handling its items, the way that these components are set up makes this really difficult to add in